src/Core/Framework/DataAbstractionLayer/Search/CompressedCriteriaDecoder.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\DataAbstractionLayer\Search;
+
+use Shopware\Core\Framework\DataAbstractionLayer\DataAbstractionLayerException;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Util\Base64;
+use Shopware\Core\Framework\Util\Exception\Base64DecodingException;
+
+/**
+ * @interal
+ */
+#[Package('framework')]
+class CompressedCriteriaDecoder
+{
+    /**
+     * Using 20x multiplier as an optimistic compression rate avoid decompression bombs.
+     */
+    private const DEFAULT_DECOMPRESSED_CRITERIA_LENGTH_LIMIT = self::DEFAULT_COMPRESSED_CRITERIA_LENGTH_LIMIT * 20;
+
+    /**
+     * Setting default to 128kb, which should be a reasonable limit for parameters provided via URL.
+     */
+    private const DEFAULT_COMPRESSED_CRITERIA_LENGTH_LIMIT = 1024 * 128;
+
+    /**
+     * @internal
+     */
+    public function __construct(
+        /**
+         * Limit for the maximum allowed length of the compressed criteria string.
+         */
+        private readonly int $compressedCriteriaLengthLimit = self::DEFAULT_COMPRESSED_CRITERIA_LENGTH_LIMIT,
+
+        /**
+         * Limit for the maximum allowed length of the decompressed criteria string to avoid decompression bombs.
+         */
+        private readonly int $decompressedCriteriaLengthLimit = self::DEFAULT_DECOMPRESSED_CRITERIA_LENGTH_LIMIT,
+    ) {
+    }
+
+    /**
+     * Decodes and decompresses the given criteria string formed through: json_encode -> gzip -> base64url_encode
+     *
+     * @throws DataAbstractionLayerException
+     *
+     * @return array<string, mixed>
+     */
+    public function decode(string $encodedCriteria): array
+    {
+        try {
+            // Hard limit to avoid overloading
+            if (\strlen($encodedCriteria) > $this->compressedCriteriaLengthLimit) {
+                throw DataAbstractionLayerException::invalidCompressedCriteriaParameter('The _criteria parameter is too long');
+            }
+
+            // Decode base64url
+            try {
+                $gzippedData = Base64::urlDecode($encodedCriteria);
+            } catch (Base64DecodingException $e) {
+                throw DataAbstractionLayerException::invalidCompressedCriteriaParameter('Unable to decode base64 data');
```

Comment:
```suggestion
                throw DataAbstractionLayerException::invalidCompressedCriteriaParameter($e->getMessage());
```

maybe like this instead? so you do not need to duplicate the message string?